File: 1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
// Time: O(n * log(n));
// Space: O(n);
var verticalTraversal = function(root) {
    if (!root) return [];
    let minLeft = 0, maxRight = 0;

    const getWidth = (nde, w) => {
        if (!nde) return;
        minLeft = Math.min(w, minLeft);
        maxRight = Math.max(w, maxRight);
        getWidth(nde.left, w - 1);
        getWidth(nde.right, w + 1);
    }
    getWidth(root, 0);

    const columns = [...new Array(Math.abs(minLeft) + maxRight + 1)].map(a => []);
    const queue = [[[root, 0]]];

    while (queue.length) {
        let level = queue.shift();
        let newLevel = [];
        for (let [nde, width] of level) {

            if (nde.left) newLevel.push([nde.left, width - 1]);
            if (nde.right) newLevel.push([nde.right, width + 1]);
        }
        level.sort((a,b) => a[0].val - b[0].val)
        for (let [nde, width ] of level) {
            columns[Math.abs(minLeft) + width].push(nde.val);
        }
        if (newLevel.length) queue.push(newLevel);
    }
    
    return columns;
};

File: 1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Data {
    var col: Int
    var node: TreeNode
    init(_ col: Int, _ node: TreeNode) { self.col = col; self.node = node }
}
class Solution {
    func verticalTraversal(_ root: TreeNode?) -> [[Int]] {
        guard var r = root else { return [] }
        var minLeft: Int = 0, maxRight: Int = 0

        func dfs(_ node: TreeNode?, _ col: Int) -> Int {
            guard var n = node else { return 0 }
            minLeft = min(minLeft, col)
            maxRight = max(maxRight, col)
            dfs(n.left, col - 1)
            dfs(n.right, col + 1)
            return col
        }
        dfs(root, 0)
        var offSet: Int = abs(minLeft)

        var width: Int = offSet + maxRight + 1
        var cols: [[Int]] = Array(repeating: [], count: width)
        var queue: [[Data]] = [[Data(0, r)]]

        while !queue.isEmpty {
            var level: [Data] = queue.removeFirst()
            var newNodes: [Data] = []
            var lvl: Int = 0
            for n: Data in level {
                cols[n.col+offSet].append(n.node.val)
                lvl = n.col+offSet
                if var left: TreeNode = n.node.left { newNodes.append(Data(n.col - 1, left))}
                if var right: TreeNode = n.node.right { newNodes.append(Data(n.col + 1, right))}
            }
            if !newNodes.isEmpty{queue.append(newNodes.sorted{$0.node.val < $1.node.val})}
           
        }
        return cols
    }
}
